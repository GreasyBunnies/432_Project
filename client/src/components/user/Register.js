import React from 'react';
import {
    Form,
    Button
} from 'react-bootstrap';
import {
    withRouter
} from 'react-router-dom';

import User from '../../model/User';
import {
    setSession
} from '../../utils/auth';

class Register extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            username: '',
            password: '',
            firstname: '',
            lastname: ''
        };

        this.handleChange = this.handleChange.bind(this);
        this.register = this.register.bind(this);
    }

    handleChange(e) {
        let name = e.target.name;
        let value = e.target.value;

        this.setState({
            [name]: value
        });
    }

    // Takes the input in the form to register the user
    register(e) {
        e.preventDefault();

        // Validate inputs
        let username = this.state.username;
        let password = this.state.password;
        let firstname = this.state.firstname;
        let lastname = this.state.lastname;

        username = username.trim();
        password = password.trim();
        firstname = firstname.trim();
        lastname = lastname.trim();

        // Fire POST request
        let url = `${process.env.REACT_APP_SERVER_DEV_DOMAIN}/api/register`;

        fetch(url, {
            method: 'POST',
            mode: 'cors',
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify({
                "username": username,
                "password": password,
                "firstname": firstname,
                "lastname": lastname
            })
        }).then(resp => {
            resp.json().then(resp => {
                let id = resp.id;

                // Instantiate User
                const user = new User(id, username, firstname, lastname, true, JSON.stringify("[]"));

                // Set user session
                setSession(user);

                // Redirect to home
                this.props.history.push("/");
            }).catch(err => {
                console.log(err);
            })
        }).catch(err => {
            console.log(err);
        })
    }

    render() {
        return (
            <div id="login">
                <Form onSubmit={this.register}>
                    <Form.Group>
                        <Form.Label>Username</Form.Label>
                        <Form.Control type="text" placeholder="Enter username" name="username" onChange={this.handleChange} />
                    </Form.Group>

                    <Form.Group>
                        <Form.Label>First Name</Form.Label>
                        <Form.Control type="text" placeholder="Enter first name" name="firstname" onChange={this.handleChange} />
                    </Form.Group>

                    <Form.Group>
                        <Form.Label>Last Name</Form.Label>
                        <Form.Control type="text" placeholder="Enter last name" name="lastname" onChange={this.handleChange} />
                    </Form.Group>

                    <Form.Group>
                        <Form.Label>Password</Form.Label>
                        <Form.Control type="password" placeholder="Password" name="password" onChange={this.handleChange} />
                    </Form.Group>

                    <Button variant="info" type="submit">
                        Register
                    </Button>
                </Form>
            </div>
        )
    }
}

export default withRouter(Register);