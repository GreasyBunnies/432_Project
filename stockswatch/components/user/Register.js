import React from 'react';
import {
	Form,
	Button
} from 'react-bootstrap';
import Router from "next/router";

import Error from '../alert/Error';
import User from '../../model/User';

import { setSession } from '../../utils/auth';
import { post } from "../../utils/requests"

class Register extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			username: '',
			password: '',
			firstname: '',
			lastname: '',
			error: false
		};

		this.handleChange = this.handleChange.bind(this);
		this.register = this.register.bind(this);
		this.toLogin = this.toLogin.bind(this);
	}

	handleChange(e) {
		let name = e.target.name;
		let value = e.target.value;

		this.setState({ [name]: value });
	}

	// Takes the input in the form to register the user
	async register(e) {
		e.preventDefault();

		// Validate inputs
		let username = this.state.username.trim().toLowerCase();
		let password = this.state.password.trim();
		let firstname = this.state.firstname.trim().toLowerCase();
		let lastname = this.state.lastname.trim().toLowerCase();

		firstname = firstname.charAt(0).toUpperCase() + firstname.slice(1);
		lastname = lastname.charAt(0).toUpperCase() + lastname.slice(1);

		// Prepare data and url
		const data = JSON.stringify({
			"username": username,
			"password": password,
			"firstname": firstname,
			"lastname": lastname
		});
		const url = `${process.env.REACT_APP_SERVER_DOMAIN}/api/user/register`;

		// Fire POST request
		try {
			const resp = await post(url, data);

			// On 200 status
			if (resp.status === 200) {
				const json = await resp.json();
				let id = json.id;

				// Instantiate User
				const user = new User(
					id,
					username,
					firstname,
					lastname,
					true,
					[]
				);

				// Set user session
				setSession(user);

				// Redirect to home
				Router.push("/");
			}

			// On 500 status
			if (resp.status === 500) this.setState({ error: true });
		} catch (err) {
			this.setState({ error: true });
		}
	}

	// Redirects user to /login
	toLogin() {
		Router.push("/login");
	}

	render() {
		return (
			<React.Fragment>
				{this.state.error ? <Error message={"There has been an error registering your information. Please try again!"} /> : ""}

				<div id="login">
					<Form onSubmit={this.register}>
						<Form.Group>
							<Form.Label>Username</Form.Label>
							<Form.Control type="text" placeholder="Enter username" name="username" onChange={this.handleChange} />
						</Form.Group>

						<Form.Group>
							<Form.Label>First Name</Form.Label>
							<Form.Control type="text" placeholder="Enter first name" name="firstname" onChange={this.handleChange} />
						</Form.Group>

						<Form.Group>
							<Form.Label>Last Name</Form.Label>
							<Form.Control type="text" placeholder="Enter last name" name="lastname" onChange={this.handleChange} />
						</Form.Group>

						<Form.Group>
							<Form.Label>Password</Form.Label>
							<Form.Control type="password" placeholder="Password" name="password" onChange={this.handleChange} />
						</Form.Group>

						<Button variant="info" type="submit">
							Register
                    	</Button>

						<span className="space"></span>

						<Button variant="outline-info" onClick={this.toLogin}>
							Login
                    	</Button>
					</Form>
				</div>
			</React.Fragment>
		)
	}
}

export default Register;