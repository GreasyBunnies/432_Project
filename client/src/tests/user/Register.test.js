import React from 'react';
import Register from '../../components/user/Register.js';
import { shallow, mount, render } from 'enzyme';

jest.mock('../__mocks__/fetchUser')

describe('Rendering Register', () => {
	it('Should render', () => {
		const component = shallow(<Register />);

		expect(component).toMatchSnapshot();
	});
});