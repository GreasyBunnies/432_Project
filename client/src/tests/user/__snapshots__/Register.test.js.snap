// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering Register Should render 1`] = `ShallowWrapper {}`;
